cmd/probe/capacityprobe.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2018 OpenEBS Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package probe
+
+import (
+	"io/ioutil"
+	"strconv"
+
+	"github.com/golang/glog"
+	"github.com/openebs/node-disk-manager/cmd/controller"
+	"github.com/openebs/node-disk-manager/pkg/util"
+)
+
+const (
+	capacityProbePriority = 3
+	capacityConfigKey     = "capacity-probe"
+)
+
+var (
+	capacityProbeName  = "capacity probe"
+	capacityProbeState = defaultEnabled
+)
+
+var capacityProbeRegister = func() {
+	ctrl := <-controller.ControllerBroadcastChannel
+	if ctrl == nil {
+		glog.Error("unable to configure", capacityProbeName)
+		return
+	}
+	if ctrl.NDMConfig != nil {
+		for _, probeConfig := range ctrl.NDMConfig.ProbeConfigs {
+			if probeConfig.Key == capacityConfigKey {
+				capacityProbeName = probeConfig.Name
+				capacityProbeState = util.CheckTruthy(probeConfig.State)
+				break
+			}
+		}
+	}
+
+	newRegistryProbe := &registerProbe{
+		priority:   capacityProbePriority,
+		name:       capacityProbeName,
+		state:      capacityProbeState,
+		pi:         newCapacityProbe(),
+		controller: ctrl,
+	}
+	newRegistryProbe.register()
+}
+
+// capacityProbe fills the capacity of the disk
+type capacityProbe struct {
+}
+
+func newCapacityProbe() *capacityProbe {
+	return &capacityProbe{}
+}
+
+// It is part of probe interface. Hence, empty implementation.
+func (cp *capacityProbe) Start() {}
+
+// FillDiskDetails updates the capacity of the disk , if the capacity is
+// not populated.
+func (cp *capacityProbe) FillDiskDetails(d *controller.DiskInfo) {
+
+	if d.Capacity != 0 {
+		return
+	}
+	sysPath := d.ProbeIdentifiers.UdevIdentifier
+	blockSize, err := strconv.ParseInt(d.NoOfBlocks, 10, 64)
+	if err != nil {
+		glog.Error("unable to parse the block size ", err)
+		return
+	}
+	b, err := ioutil.ReadFile(sysPath + "/queue/hw_sector_size")
+	if err != nil {
+		glog.Error("unable to read sector size", err)
+		return
+	}
+	sectorSize, err := strconv.ParseInt(string(b), 10, 64)
+	if err != nil {
+		glog.Error("unable to parse the sector size", err)
+		return
+	}
+	d.Capacity = uint64(blockSize * sectorSize)
```

Comment:
check if logical sector size is not set then set it with `sectorSize`

Reply:
@shovanmaity could you explain this?

Reply:
We are not getting logical sector size of disk also using `udev` and `smart` probe for some cases. This probe is going to fetch that by reading system file to get capacity of disk, it should update `logical sector size` also. It is not a part of issue #83 but I will suggest you to update that value if it is not set. You need to check if [logical sector size](https://github.com/openebs/node-disk-manager/blob/4655867f989bed4a56c9c22966ac8907c960de22/cmd/controller/disk.go#L42) attribute is not set then set it using this probe. (same as `d.Capacity`)